File: DictionaryParsing/python/dictionary_parser.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

def _parse_key_value_pair(pair : str) -> tuple:
    if not "=" in pair:
        raise KeyError("Equals sign is missing from input string!")
        
    key_value_pair = pair.split("=", 1)
    key = key_value_pair[0].strip()
    value = key_value_pair[1].strip()
    
    if not key:
        raise KeyError("Key is missing from input string!")
        
    return (key, value)
    

def parse(text: str) -> dict:
    
    dictionary = dict()
    
    pairs = [pair.strip() for pair in text.split(";")]
    pairs_with_content = [pair for pair in pairs if pair]
    key_value_pairs = [_parse_key_value_pair(pair) for pair in pairs_with_content]
    
    for key_value_pair in key_value_pairs:
        key, value = key_value_pair
        dictionary[key] = value
    
    return dictionary

File: DictionaryParsing/c_sharp/DictionaryParser/DictionaryParser.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DictionaryParser
{
    public static class DictionaryParser
    {
        public static Dictionary<string, string> Parse(string text)
        {
            return text.Split(';')
                .Where(pair => !string.IsNullOrWhiteSpace(pair))
                .Select(ParseKeyValuePair)
                .Aggregate(
                    new Dictionary<string, string>(),
                    AddKeyValuePairToDictionary);
        }

        private static (string key, string value) ParseKeyValuePair(string pair)
        {
            if (!pair.Contains('='))
            {
                throw new ArgumentException("No equals sign in pair");
            }

            var split = pair.Split('=', 2);
            var key = split[0].Trim();
            var value = split[1].Trim();

            if (string.IsNullOrEmpty(key))
            {
                throw new ArgumentException("No key in pair");
            }

            return (key, value);
        }

        private static Dictionary<string, string> AddKeyValuePairToDictionary(
            Dictionary<string, string> dictionary,
            (string key, string value) keyValuePair)
        {
            var (key, value) = keyValuePair;
            dictionary[key] = value;
            return dictionary;
        }
    }
}
